bindings/rust/evmc-client/src/host.rs

```diff
@@ -0,0 +1,315 @@
+/* EVMC: Ethereum Client-VM Connector API.
+ * Copyright 2019 The EVMC Authors.
+ * Licensed under the Apache License, Version 2.0.
+ */
+
+use crate::types::*;
+use evmc_sys as ffi;
+use std::mem;
+
+pub trait HostInterface {
+    fn account_exists(&mut self, addr: &Address) -> bool;
+    fn get_storage(&mut self, addr: &Address, key: &Bytes32) -> Bytes32;
+    fn set_storage(&mut self, addr: &Address, key: &Bytes32, value: &Bytes32) -> StorageStatus;
+    fn get_balance(&mut self, addr: &Address) -> Bytes32;
+    fn get_code_size(&mut self, addr: &Address) -> usize;
+    fn get_code_hash(&mut self, addr: &Address) -> Bytes32;
+    fn copy_code(
+        &mut self,
+        addr: &Address,
+        offset: &usize,
+        buffer_data: &*mut u8,
+        buffer_size: &usize,
+    ) -> usize;
+    fn selfdestruct(&mut self, addr: &Address, beneficiary: &Address);
+    fn get_tx_context(&mut self) -> (Bytes32, Address, Address, i64, i64, i64, Bytes32, Bytes32);
+    fn get_block_hash(&mut self, number: i64) -> Bytes32;
+    fn emit_log(&mut self, addr: &Address, topics: &Vec<Bytes32>, data: &Bytes);
+    fn call(
+        &mut self,
+        kind: CallKind,
+        destination: &Address,
+        sender: &Address,
+        value: &Bytes32,
+        input: &Bytes,
+        gas: i64,
+        depth: i32,
+        is_static: bool,
+    ) -> (Vec<u8>, i64, Address, StatusCode);
+}
+
+struct Callback {
+    host_interface: ::std::option::Option<Box<dyn HostInterface>>,
+}
+
+static mut CALLBACK: Callback = Callback {
```

Comment:
This makes this not thread-safe. Why did you choose to use this and not use `evmc_host_context`, which by design is there to avoid having these globals.

Reply:
Yes, you are right. The original implementation was limited by how much I familiar with Rust. I have rewritten it more like go-binding implementation in 5e150ea.